### ISSUE TITLE
Timeout errors during stress load on a cluster while it's doing repair

### ISSUE THREAD

**Issue by noamha** [scylladb/scylla#861]
Configuration: 3 x Rackspace OnMetal IO v1, Scylla 0.17, RF x 2
Init Keyspace

```
cassandra-stress write n=1 cl=ALL -schema "replication(factor=2)" -mode native cql3 -node $SERVERS
```

Wrote 150M keys using 5 stress servers:

```
time cassandra-stress write cl=TWO n=30000000 -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=1..30000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress write cl=TWO n=30000000 -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=30000000..60000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress write cl=TWO n=30000000 -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=60000000..90000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress write cl=TWO n=30000000 -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=90000000..120000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress write cl=TWO n=30000000 -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=120000000..150000000 -col 'size=FIXED(2048) n=FIXED(5)
```
- After the stress each node had 1100 GB of data, I've stopped node3 and reduce it's data to 950GB by deleting random SSTables.
- Started running repair on the node3
- While repair is running started also running the following stress from 5 stress servers, 2 writers and 3 readers:

```
time cassandra-stress write cl=TWO duration=5h -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=150000000..180000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress write cl=TWO duration=5h -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=180000000..210000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress read cl=TWO duration=5h -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=1..50000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress read cl=TWO duration=5h -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=50000000..100000000 -col 'size=FIXED(2048) n=FIXED(5)'
time cassandra-stress read cl=TWO duration=5h -mode native cql3 -rate threads=700 -node $SERVERS -pop seq=100000000..150000000 -col 'size=FIXED(2048) n=FIXED(5)'
```
- Got a lot of timeout errors from stress severs:

```
com.datastax.driver.core.exceptions.WriteTimeoutException: Cassandra timeout during write query at consistency TWO (2 replica were required but only 1 acknowledged the write)
```
- I've stopped of all the stress servers and ran from only a single write stress server, and got a lot of the following errors at about 6K op/s:

```
com.datastax.driver.core.exceptions.WriteTimeoutException: Cassandra timeout during write query at consistency TWO (2 replica were required but only 1 acknowledged the write)
```
- I've limit the stress write server to 5K op/s and things seems to stable
- After a few minutes I started putting more stress on the cluster by using more stress servers and removing the limits, it seems the cluster was able to handle more load, but still threw exceptions sometimes

**Comment by dorlaor** [comment]
Seems like we need larger involvement of the I/O scheduler in our repair

**Comment by nyh** [comment]
On Mon, Feb 1, 2016 at 4:10 PM, Dor Laor <notifications@github.com> wrote:

> Seems like we need larger involvement of the I/O scheduler in our repair
>

Indeed. Also note that repair is very parallel - not only do we split the
token space to around 256 ranges and repair them in parallel, we also may
send the other host as many as 100 checksum requests (for each one of these
256 ranges), and there could be many of these other hosts - all in
parallel. So indeed the I/O scheduler should limit the disk bandwidth that
all this parallel things are getting. Since repair has almost no
"computational" work beyond reading from disk and parsing it - limiting the
I/O would be enough to limit everything else, I think.

**Comment by avikivity** [comment]
The default I/O scheduler configuration doesn't limit anything.

**Comment by dorlaor** [comment]
On Mon, Feb 1, 2016 at 4:32 PM, Avi Kivity <notifications@github.com> wrote:

> The default I/O scheduler configuration doesn't limit anything.
>

Sounds like a real low hanging fruit for this infrastructure

**Comment by avikivity** [comment]
I'll be happy to help tune it, but we need to ship a tool to help users.  @glommer

**Comment by glommer** [comment]
One very easy thing we could do is to limit by 128 * nr_disks. This is an upper limit of what Linux can handle anyway (in a default configuration), and issuing more than that will block.

This value is not 100 % accurate, because in reality with more than 1 disk requests can pile up in one of them and it seems to me that the RAID code will not necessarily try to stop that condition from happening.

Right now we have 128 * nr_shards, which is basically a way to say "disable", since even fast disks won't reach this limit. So my only concern is how would setting it to 128 * nr_disks would limit our parallelism for fast scenarios. Maybe it wouldn't hurt much, since fast disks will also tend to complete those requests quite fast - and again, if we do push more than 128 and they pile up in the kernel, it will block anyway.

For high-profile public infrastructure like Amazon, Rackspace, GCE, etc, a better solution could possibly be to just run diskplorer ourselves and come up with recommendations (done for Amazon already). It is certainly more efficient than having each user do that by himself, better than blind defaults, and only worse than fully automatic tuning.

**Comment by dorlaor** [comment]
@noamha will be thrilled to have the honour of running the notorious
diskplorer on Rackspace

**Comment by avikivity** [comment]
128*nr_disks is very inaccurate, considering that some SATA disks can handle <10 concurrent requests.

Can we do something like run fio in the same way as diskplorer, but instead of displaying the graph, take concurrency @ 80% max bandwidth and multiply by 3?  Then divide into io_queues so that each queue has at a minimum concurrency of 4.

**Comment by glommer** [comment]
On Mon, Feb 1, 2016 at 11:01 AM, Avi Kivity <notifications@github.com>
wrote:

> 128*nr_disks is very inaccurate, considering that some SATA disks can
> handle <10 concurrent requests.
>
> Can we do something like run fio in the same way as diskplorer, but
> instead of displaying the graph, take concurrency @ 80% max bandwidth and
> multiply by 3? Then divide into io_queues so that each queue has at a
> minimum concurrency of 4.
>

Displaying or not displaying the graph, this is a very time consuming
process that I bet most people would either not do at all, or just do it
with the strong preference towards not having to have had done it.
Specially because we will still be some eons away from fully automating all
that. When we can fully automate it, that's (slightly) another story.

For public clouds (be it virtual or baremetal) where machines will tend to
have similar characteristics, I would still say that the most efficient way
is to have somebody (us) to it once and come up with good recommendations
for different machine classes.

Another option of course is to just use default values based on the amount
of disks as I have suggested, but augmenting that information with
disk-type information, assuming we can safely get that from sysfs. So once
we see a bus+disk class, we can then determine R so that max_nr_requests is
set to R * nr_disks. What do you think of that ?

**Comment by avikivity** [comment]
We haven't seen enough disks to have confidence in static classification,
IMO.

If we code some sort of binary search, we may be able to converge in around
a minute. The program will then write some conf file which the init script
can read.

**Comment by slivne** [comment]
Lets use the diskplorer on rackspace and get numbers for it and set the values manually to verify that indeed by setting values this will be fixed.

**Comment by dorlaor** [comment]
Alon has a running setup, someone (@slivne?) will need to guide him what to
run

**Comment by nyh** [comment]
I just sent a patch for #887 to the mailing list. Would be great if someone who knows how to repeat this use case can verify that it makes this problem go away.

The theory is that with this patch, despite repair being so parallelized, it will get an equal share of the disk bandwidth and requests can still continue - at a slower pace but never get hugely backlogged.

Moreover, this patch also gives you control on how to make repair use even less disk bandwidth if you want to slow it down even more (and give more performance to the read requests): in priority_manager.hh, you can change the "100" in the line
```
_repair_priority(engine().register_one_priority_class("repair", 100))
```
To a lower number if you want repair to get a smaller share of the disk bandwidth.

**Comment by tzach** [comment]
@glommer is this fix?

**Comment by glommer** [comment]
Should be.

There are still outstanding problems but they will only happen if we have
more than one CF being repaired and we have a separate ussue for that one.

**Comment by noamha** [comment]
Ran a similar test as before, only with the following exception:
- Scylla 1.5 AMI
- i2.8xlarge
And all stress test ran with 100 threads instead of 700 threads.

Didn't get any errors while running the repair.
